flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/nn/KNN.scala

```diff
@@ -0,0 +1,354 @@
+                for ((training, testing) <- iter) {
+                  val queue = mutable.PriorityQueue[(FlinkVector, FlinkVector, Long, Double)]()(
+                    Ordering.by(_._4))
+
+                  // use a quadtree if (4^dim)Ntest*log(Ntrain)
+                  // < Ntest*Ntrain, and distance is Euclidean
+                  val useQuadTree = resultParameters.get(UseQuadTreeParam).getOrElse(
+                    training.values.head.size + math.log(math.log(training.values.length) /
+                      math.log(4.0)) < math.log(training.values.length) / math.log(4.0) &&
+                      (metric.isInstanceOf[EuclideanDistanceMetric] ||
+                        metric.isInstanceOf[SquaredEuclideanDistanceMetric]))
+
+                  if (useQuadTree) {
+                    knnQueryWithQuadTree(training.values, testing.values, k, metric, queue, out)
+                  } else {
+                    knnQueryBasic(training.values, testing.values, k, metric, queue, out)
+                  }
+                }
+              }
+            }
+
+            // group by input vector id and pick k nearest neighbor for each group
+            val result = crossed.groupBy(2).sortGroup(3, Order.ASCENDING).reduceGroup {
+              (iter, out: Collector[(FlinkVector, Array[FlinkVector])]) => {
+                if (iter.hasNext) {
+                  val head = iter.next()
+                  val key = head._2
+                  val neighbors: ArrayBuffer[FlinkVector] = ArrayBuffer(head._1)
+
+                  for ((vector, _, _, _) <- iter.take(k - 1)) {
+                    // we already took a first element
+                    neighbors += vector
+                  }
+
+                  out.collect(key, neighbors.toArray)
+                }
+              }
+            }
+
+            result
+          case None => throw new RuntimeException("The KNN model has not been trained." +
+            "Call first fit before calling the predict operation.")
+
+        }
+      }
+    }
+  }
+
+  def knnQueryWithQuadTree[T <: FlinkVector](
+    training: Vector[T],
+    testing: Vector[(Long, T)],
+    k: Int, metric: DistanceMetric,
+    queue: mutable.PriorityQueue[(FlinkVector,
+      FlinkVector, Long, Double)],
+    out: Collector[(FlinkVector,
+      FlinkVector, Long, Double)]) {
+    /// find a bounding box
+    val MinArr = Array.tabulate(training.head.size)(x => x)
+    val MaxArr = Array.tabulate(training.head.size)(x => x)
+
+    val minVecTrain = MinArr.map(i => training.map(x => x(i)).min - 0.01)
+    val minVecTest = MinArr.map(i => testing.map(x => x._2(i)).min - 0.01)
+    val maxVecTrain = MaxArr.map(i => training.map(x => x(i)).max + 0.01)
+    val maxVecTest = MaxArr.map(i => testing.map(x => x._2(i)).max + 0.01)
+
+    val MinVec = DenseVector(MinArr.map(i => Array(minVecTrain(i), minVecTest(i)).min))
+    val MaxVec = DenseVector(MinArr.map(i => Array(maxVecTrain(i), maxVecTest(i)).max))
```

Comment:
Agreed.  Only changed `Array(minVecTrain(i), minVecTest(i)).min` to `math.min(minVecTrain(i), minVecTest(i))` and similarly for max (assume there was no more efficient way for the preceeding lines involving min/max)